File: Programmers/고득점Kit/타겟넘버_DFS.py
# 재귀를 이용한 DFS(backtracking)
def solution(numbers, target):
    answer = 0 # 전역변수가 아닌 지역변수
    n = len(numbers)
    
    def dfs(idx, result):
        if idx == n:
            if result == target:
                nonlocal answer # 지역변수를 전역변수처럼 사용
                answer += 1
            return # 마지막 차수까지 도달했으므로 리턴
        else:
            dfs(idx + 1, result + numbers[idx]) # 먼저 더하기를 진행한 다음
            dfs(idx + 1, result - numbers[idx]) # return 되어 나왔을 때 빼기 진행 (backtrack)
            
    dfs(0, 0)
    
    return answer

File: Programmers/고득점Kit/타겟넘버_DFS.java
class Solution {
    
    static int count;
    
    public int solution(int[] numbers, int target) {
        
        count = 0;
        
        DFS(numbers, target, 0, 0);
        
        return count;
    
    }
    
    public int DFS(int[] numbers, int target, int index, int sum) {
        
        if (index == numbers.length && sum == target) {
            count++;
            return 0;
        }
        
        if (index == numbers.length) {
            return 0;
        }
        
        DFS(numbers, target, index + 1, sum + numbers[index]);
        DFS(numbers, target, index + 1, sum - numbers[index]);
        
        return 0;
        
    }
               
}